File: frontend/src/utils/storage.js
import { requestJson, request } from '../api/client'
import { getSpotifyTrackId } from './spotify'

const SEARCH_CACHE_TTL_MS = 7 * 24 * 60 * 60 * 1000

const PICK_MODES = ['quality', 'speed', 'longest', 'balanced']
const FORMAT_PREFS = ['any', 'flac', 'mp3', 'ogg', 'm4a']

// ---- history ---------------------------------------------------------------
export function getHistoryLabel(url, index) {
  try {
    const parsed = new URL(url)
    const parts = parsed.pathname.split('/').filter(Boolean)
    const id = parts.at(-1)
    return id ? `Playlist ${index + 1} · ${id}` : `Playlist guardada ${index + 1}`
  } catch {
    return `Playlist guardada ${index + 1}`
  }
}

export function normalizeHistory(entries) {
  return (Array.isArray(entries) ? entries : [])
    .map((entry, index) => {
      const item = typeof entry === 'string' ? { url: entry } : entry
      if (!item?.url) return null
      return {
        url: item.url,
        name: item.name || item.title || getHistoryLabel(item.url, index),
      }
    })
    .filter(Boolean)
}

export async function loadHistory() {
  try {
    return normalizeHistory(await requestJson('/api/storage/history'))
  } catch {
    return []
  }
}

export async function saveHistory(history) {
  try {
    await request('/api/storage/history', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ entries: normalizeHistory(history).slice(0, 20) }),
    })
  } catch {}
}

export async function clearHistory() {
  try {
    await request('/api/storage/history', { method: 'DELETE' })
  } catch {}
}

// ---- search preferences ----------------------------------------------------
export async function loadSearchPreferences() {
  try {
    const saved = await requestJson('/api/storage/search-preferences')
    return {
      pickMode: PICK_MODES.includes(saved?.pickMode) ? saved.pickMode : 'quality',
      formatPref: FORMAT_PREFS.includes(saved?.formatPref) ? saved.formatPref : 'any',
    }
  } catch {
    return { pickMode: 'quality', formatPref: 'any' }
  }
}

export async function saveSearchPreferences(pickMode, formatPref) {
  try {
    await request('/api/storage/search-preferences', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ pickMode, formatPref }),
    })
  } catch {}
}

// ---- last playlist ---------------------------------------------------------
export async function loadLastPlaylist() {
  try {
    return await requestJson('/api/storage/playlist')
  } catch {
    return {}
  }
}

export async function saveLastPlaylist(playlist) {
  try {
    await request('/api/storage/playlist', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(playlist),
    })
  } catch {}
}

// ---- output folder preferences ---------------------------------------------
export async function getOutputFolderPreference(url) {
  if (!url) return undefined
  try {
    const data = await requestJson(`/api/storage/output-folder?playlist_url=${encodeURIComponent(url)}`)
    return data.folderName ?? undefined
  } catch {
    return undefined
  }
}

export async function setOutputFolderPreference(url, folderName) {
  if (!url) return
  try {
    await request('/api/storage/output-folder', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ playlist_url: url, folder_name: folderName || '' }),
    })
  } catch {}
}

// ---- search cache ----------------------------------------------------------
export async function loadSearchCache(url, tracks) {
  if (!url) return []
  try {
    const data = await requestJson(`/api/storage/search-cache?playlist_url=${encodeURIComponent(url)}`)
    const searches = data.searches || []
    return searches
      .map((search) => {
        const trackIndex = tracks.findIndex((track) =>
          (search.trackKey && getSpotifyTrackId(track.spotify_url) === search.trackKey) ||
          track.search_query === search.query
        )
        return trackIndex < 0
          ? null
          : {
              ...search,
              trackIndex,
              searchId: null,
              status: 'completed',
              cached: true,
            }
      })
      .filter(Boolean)
  } catch {
    return []
  }
}

export async function saveSearchCache(url, searches) {
  if (!url) return
  try {
    const cacheable = searches
      .filter((search) => search.raw)
      .map((search) => ({
        ...search,
        raw: {
          ...search.raw,
          results: Array.isArray(search.raw.results) ? search.raw.results.slice(0, 100) : [],
        },
      }))
    await request('/api/storage/search-cache', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ playlist_url: url, searches: cacheable }),
    })
  } catch {}
}

File: backend/routes/storage.py
"""Storage endpoints backed by SQLite (replaces localStorage)."""

from __future__ import annotations

from typing import TYPE_CHECKING

from flask import Blueprint, jsonify, request

from backend.database import (
    clear_url_history,
    get_output_folder_pref,
    load_last_playlist,
    load_search_cache,
    load_search_prefs,
    load_url_history,
    save_last_playlist,
    save_search_cache,
    save_search_prefs,
    save_url_history,
    set_output_folder_pref,
)

if TYPE_CHECKING:
    from backend.runtime import RuntimeState


def create_blueprint(state: RuntimeState) -> Blueprint:
    bp = Blueprint("storage", __name__)

    # ---- last playlist ----
    @bp.route("/api/storage/playlist", methods=["GET", "PUT"])
    def api_last_playlist():
        if request.method == "GET":
            return jsonify(load_last_playlist())
        data = request.get_json() or {}
        save_last_playlist(data)
        return jsonify({"ok": True})

    # ---- url history ----
    @bp.route("/api/storage/history", methods=["GET", "PUT", "DELETE"])
    def api_url_history():
        if request.method == "GET":
            return jsonify(load_url_history())
        if request.method == "DELETE":
            clear_url_history()
            return jsonify({"ok": True})
        data = request.get_json() or {}
        entries = data.get("entries", [])
        save_url_history(entries)
        return jsonify({"ok": True})

    # ---- search cache ----
    @bp.route("/api/storage/search-cache", methods=["GET", "PUT"])
    def api_search_cache():
        playlist_url = request.args.get("playlist_url") if request.method == "GET" else None
        if request.method == "GET":
            if not playlist_url:
                return jsonify({"searches": []})
            return jsonify({"searches": load_search_cache(playlist_url)})
        data = request.get_json() or {}
        playlist_url = data.get("playlist_url", "")
        searches = data.get("searches", [])
        if not playlist_url:
            return jsonify({"error": "Falta playlist_url"}), 400
        save_search_cache(playlist_url, searches)
        return jsonify({"ok": True})

    # ---- output folder preferences ----
    @bp.route("/api/storage/output-folder", methods=["GET", "PUT"])
    def api_output_folder():
        playlist_url = request.args.get("playlist_url") if request.method == "GET" else None
        if request.method == "GET":
            if not playlist_url:
                return jsonify({"folderName": None})
            return jsonify({"folderName": get_output_folder_pref(playlist_url)})
        data = request.get_json() or {}
        playlist_url = data.get("playlist_url", "")
        folder_name = data.get("folder_name", "")
        if not playlist_url:
            return jsonify({"error": "Falta playlist_url"}), 400
        set_output_folder_pref(playlist_url, folder_name)
        return jsonify({"ok": True})

    # ---- search preferences ----
    @bp.route("/api/storage/search-preferences", methods=["GET", "PUT"])
    def api_search_prefs():
        if request.method == "GET":
            return jsonify(load_search_prefs())
        data = request.get_json() or {}
        pick_mode = data.get("pickMode", "quality")
        format_pref = data.get("formatPref", "any")
        save_search_prefs(pick_mode, format_pref)
        return jsonify({"ok": True})

    return bp
